src/Shared/DiagnosticSourceInstrumentation/PropertyFetcher.cs

```diff
@@ -77,61 +68,119 @@ public bool TryFetch(object obj, out T value, bool skipObjNullCheck = false)
+                {
+                    return (PropertyFetch)typeof(PropertyFetch)
+                        .GetMethod(nameof(CreateInstantiated), BindingFlags.NonPublic | BindingFlags.Static)!
+                        .MakeGenericMethod(declaringType) // This is validated in the earlier call chain to be a reference type.
+                        .Invoke(null, new object[] { propertyInfo })!;
+                }
             }
         }
 
         public virtual bool TryFetch(object obj, out T value)
         {
-            value = default;
+            value = default!;
             return false;
         }
 
-        private sealed class TypedPropertyFetch<TDeclaredObject, TDeclaredProperty> : PropertyFetch
-            where TDeclaredProperty : T
+        // Goal: make PropertyFetcher AOT-compatible.
+        // AOT compiler can't guarantee correctness when call into MakeGenericType or MakeGenericMethod
+        // if one of the generic parameters is a value type (reference types are OK.)
+        // For PropertyFetcher, the decision was made to only support reference type payloads, i.e.:
+        // the object from which to get the property value MUST be a reference type.
+        // Create generics with the declared object type as a generic parameter is OK, but we need the return type
+        // of the property to be a value type (on top of reference types.)
+        // Normally, we would have a helper class like `PropertyFetchInstantiated` that takes 2 generic parameters,
+        // the declared object type, and the type of the property value.
+        // But that would mean calling MakeGenericType, with value type parameters which AOT won't support.
+        //
+        // As a workaround, Generic instantiation was split into:
+        // 1. The property value type comes from the PropertyFetcher generic parameter.
+        //    Compiler supports it even if it is a value type; the type is known statically during compilation
+        //    since PropertyFetcher is used with it.
+        // 2. Then, the declared object type was passed as a generic parameter to a generic method on PropertyFetcher<T> (or nested type.)
+        //    Therefore, calling into MakeGenericMethod will only require specifying one parameter - the declared object type.
+        //    The declared object type is guaranteed to be a reference type (throw on value type.) Thus, MakeGenericMethod is AOT compatible.
+        private static PropertyFetch CreateInstantiated<TDeclaredObject>(PropertyInfo propertyInfo)
+            where TDeclaredObject : class
+            => new PropertyFetchInstantiated<TDeclaredObject>(propertyInfo);
+
+        // ReferenceTypePropertyFetch is the optimized version because it uses CreateDelegate to get a Delegate directly to get the property.
+#if NET6_0_OR_GREATER
+        [UnconditionalSuppressMessage("AOT", "IL2109", Justification = "The code guarantees that TDeclaredObject is a reference type.")]
```

Comment:
Cleaner solution would be to put `RequiresUnreferencedCode` on this type as well.